chain-tx-enclave/tx-query-next/enclave-app/src/main.rs

```diff
@@ -1,42 +1,157 @@
+                        log::error!(
+                            "Error while writing encryption response back to TLS stream: {}",
+                            err
+                        );
+                    }
+                }
+                Ok(TxQueryInitRequest::DecryptChallenge) => {
+                    let challenge = get_random_challenge();
+
+                    if let Err(err) = stream.write_all(&challenge) {
+                        log::error!("Unable to write random challenge to TLS stream: {}", err);
+                        return;
                     }
+
+                    match stream.read(&mut bytes) {
+                        Ok(len) => {
+                            match DecryptionRequest::decode(&mut &bytes.as_slice()[0..len]) {
+                                Ok(decryption_request) => {
+                                    if !verify_decryption_request(&decryption_request, challenge) {
+                                        log::error!("Decryption request is invalid");
+                                        return;
+                                    }
+
+                                    match handle_decryption_request(&decryption_request, zmq_stream)
+                                    {
+                                        Ok(decryption_response) => {
+                                            if let Err(err) =
+                                                stream.write_all(&decryption_response.encode())
+                                            {
+                                                log::error!("Error while writing decryption response back to TLS stream: {}", err);
+                                            }
+                                        }
+                                        Err(err) => log::error!(
+                                            "Error while handling decryption request: {}",
+                                            err
+                                        ),
+                                    }
+                                }
+                                Err(err) => {
+                                    log::error!("Unable to decode decryption request: {}", err)
+                                }
+                            }
+                        }
+                        Err(err) => {
+                            log::error!(
+                                "Unable to read challenge response from TLS stream: {}",
+                                err
+                            );
+                        }
+                    }
+                }
+                Err(err) => {
+                    log::error!("Error while decoding tx-query init request: {}", err);
+                    return;
                 }
-            }
-            Err(_) => {}
+            };
         }
+        Err(err) => log::error!("Error while reading bytes from TLS stream: {}", err),
```

Comment:
this is pretty complex and hard to read -- and shouldn't some of the errors at least be written back to the stream and/or send TLS `close_notify`?

Reply:
Ideally, there should be a way to write error messages back to the stream. But, we'll have to modify `enclave-protocol` to do that as there is not way to send back error messages in current types.

Reply:
and `.send_close_notify()` ?

Reply:
`.send_close_notify()` will not add any value here because the `TcpStream` gets dropped right after returning from this function (which closes the connection) and we don't write any additional message to the stream before closing connection.

Reply:
it'd be more graceful?

Reply:
How? `.send_close_notify()` will only send something to the client when we write on the stream ([docs](https://docs.rs/rustls/0.17.0/rustls/trait.Session.html#tymethod.send_close_notify)).

Reply:
+ empty write?